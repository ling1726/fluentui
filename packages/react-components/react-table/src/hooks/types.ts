import { SortDirection } from '../components/Table/Table.types';
import { TableHeaderCellProps } from '../components/TableHeaderCell/TableHeaderCell.types';

export type RowId = string | number;
export type ColumnId = string | number;
export type GetRowIdInternal<TItem> = (rowId: TItem, index: number) => RowId;

export interface ColumnDefinition<TItem> {
  columnId: ColumnId;
  compare?: (a: TItem, b: TItem) => number;
}

export interface SortStateInternal<TItem> {
  sortDirection: SortDirection;
  sortColumn: ColumnId | undefined;
  toggleSort: (columnId: ColumnId) => void;
  headerSortProps: (columnId: ColumnId) => TableHeaderCellProps;
  /**
   * Returns a sorted **shallow** copy of original items
   */
  sort: (items: TItem[]) => TItem[];
}

export interface UseTableOptions<TItem> {
  columns: ColumnDefinition<TItem>[];
  items: TItem[];
  selectionMode?: 'single' | 'multiselect';
  getRowId?: (item: TItem) => RowId;
}

export interface SelectionStateInternal {
  clearSelection: () => void;
  deSelectRow: (rowId: RowId) => void;
  selectRow: (rowId: RowId) => void;
  toggleSelectAllRows: () => void;
  toggleRowSelect: (rowId: RowId) => void;
  selectedRows: Set<RowId>;
  allRowsSelected: boolean;
  someRowsSelected: boolean;
}

export interface SortState {
  sortDirection: SortDirection;
  sortColumn: ColumnId | undefined;
  toggleSort: (columnId: ColumnId) => void;
  headerSortProps: (columnId: ColumnId) => TableHeaderCellProps;
}

export interface SelectionState {
  /**
   * Clears all selected rows
   */
  clearSelection: () => void;
  /**
   * Selects single row
   */
  selectRow: (rowId: RowId) => void;
  /**
   * De-selects single row
   */
  deSelectRow: (rowId: RowId) => void;
  /**
   * Toggle selection of all rows
   */
  toggleSelectAllRows: () => void;
  /**
   * Toggle selection of single row
   */
  toggleRowSelect: (rowId: RowId) => void;
  /**
   * Collection of row ids corresponding to selected rows
   */
  selectedRows: RowId[];
  /**
   * Whether all rows are selected
   */
  allRowsSelected: boolean;
  /**
   * Whether some rows are selected
   */
  someRowsSelected: boolean;
}

export interface RowState<TItem> {
  /**
   * User provided data
   */
  item: TItem;
  /**
   * Toggle the selection of the row
   */
  toggleSelect: () => void;
  /**
   * Selects the row
   */
  selectRow: () => void;
  /**
   * De-selects the row
   */
  deSelectRow: () => void;
  /**
   * Whether the row is selected
   */
  selected: boolean;
  /**
   * The row id, defaults to index position in the collection
   */
  rowId: RowId;
}

export interface TableState<TItem> {
  /**
   * Gets the row data for rendering
   */
  rows: () => RowState<TItem>[];
  /**
   * State and actions to manage row selection
   */
  selection: SelectionState;
  /**
   * State and actions to manage row sorting
   */
  sort: SortState;
}
