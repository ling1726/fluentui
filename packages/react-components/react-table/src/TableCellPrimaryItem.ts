export * from './components/TableCellPrimaryItem/index';
