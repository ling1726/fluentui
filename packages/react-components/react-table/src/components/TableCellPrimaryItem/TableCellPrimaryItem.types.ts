import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import { TableCellItemSlots } from '../TableCellItem';

export type TableCellPrimaryItemSlots = {
  main: Slot<'span'>;

  secondary: Slot<'span'>;

  wrapper: Slot<'div'>;
} & TableCellItemSlots;

/**
 * TableCellPrimaryItem Props
 */
export type TableCellPrimaryItemProps = ComponentProps<Partial<TableCellPrimaryItemSlots>> & {};

/**
 * State used in rendering TableCellPrimaryItem
 */
export type TableCellPrimaryItemState = ComponentState<TableCellPrimaryItemSlots>;
