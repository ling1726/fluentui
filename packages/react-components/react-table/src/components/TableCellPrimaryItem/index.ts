export * from './TableCellPrimaryItem';
export * from './TableCellPrimaryItem.types';
export * from './renderTableCellPrimaryItem';
export * from './useTableCellPrimaryItem';
export * from './useTableCellPrimaryItemStyles';
