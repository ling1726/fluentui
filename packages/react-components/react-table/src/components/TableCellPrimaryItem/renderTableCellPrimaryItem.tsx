import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TableCellPrimaryItemState, TableCellPrimaryItemSlots } from './TableCellPrimaryItem.types';

/**
 * Render the final JSX of TableCellPrimaryItem
 */
export const renderTableCellPrimaryItem_unstable = (state: TableCellPrimaryItemState) => {
  const { slots, slotProps } = getSlots<TableCellPrimaryItemSlots>(state);

  return (
    <slots.root {...slotProps.root}>
      {slots.media && <slots.media {...slotProps.media} />}
      {slots.wrapper && (
        <slots.wrapper {...slotProps.wrapper}>
          {slots.main && <slots.main {...slotProps.main} />}
          {slots.secondary && <slots.secondary {...slotProps.secondary} />}
        </slots.wrapper>
      )}
      {slotProps.root.children}
    </slots.root>
  );
};
