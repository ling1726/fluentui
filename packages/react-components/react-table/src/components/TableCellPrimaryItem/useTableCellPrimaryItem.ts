import * as React from 'react';
import { resolveShorthand } from '@fluentui/react-utilities';
import type { TableCellPrimaryItemProps, TableCellPrimaryItemState } from './TableCellPrimaryItem.types';
import { useTableCellItem_unstable } from '../TableCellItem/useTableCellItem';

/**
 * Create the state required to render TableCellPrimaryItem.
 *
 * The returned state can be modified with hooks such as useTableCellPrimaryItemStyles_unstable,
 * before being passed to renderTableCellPrimaryItem_unstable.
 *
 * @param props - props from this instance of TableCellPrimaryItem
 * @param ref - reference to root HTMLElement of TableCellPrimaryItem
 */
export const useTableCellPrimaryItem_unstable = (
  props: TableCellPrimaryItemProps,
  ref: React.Ref<HTMLElement>,
): TableCellPrimaryItemState => {
  const tableCellState = useTableCellItem_unstable(props, ref);

  return {
    ...tableCellState,
    components: {
      ...tableCellState.components,
      main: 'span',
      secondary: 'span',
      wrapper: 'div',
      media: 'span',
    },
    main: resolveShorthand(props.secondary),
    secondary: resolveShorthand(props.main),
    wrapper: resolveShorthand(props.wrapper, { required: !!props.main || !!props.secondary }),
  };
};
