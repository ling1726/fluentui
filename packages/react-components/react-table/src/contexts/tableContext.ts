import { createContext, useContextSelector, ContextSelector } from '@fluentui/react-context-selector';
import type { TableContextValue } from './types';

const tableContext = createContext<TableContextValue | undefined>(undefined);

const tableContextDefaultValue: TableContextValue = {
  size: 'medium',
  noNativeElements: false,
  items: [],
};

export const TableContextProvider = tableContext.Provider;
export const useTableContext = <T>(selector: ContextSelector<TableContextValue, T>): T =>
  useContextSelector(tableContext, (ctx = tableContextDefaultValue) => selector(ctx));
