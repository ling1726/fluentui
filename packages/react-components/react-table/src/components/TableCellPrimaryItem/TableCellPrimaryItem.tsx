import * as React from 'react';
import { useTableCellPrimaryItem_unstable } from './useTableCellPrimaryItem';
import { renderTableCellPrimaryItem_unstable } from './renderTableCellPrimaryItem';
import { useTableCellPrimaryItemStyles_unstable } from './useTableCellPrimaryItemStyles';
import type { TableCellPrimaryItemProps } from './TableCellPrimaryItem.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * TableCellPrimaryItem component - TODO: add more docs
 */
export const TableCellPrimaryItem: ForwardRefComponent<TableCellPrimaryItemProps> = React.forwardRef((props, ref) => {
  const state = useTableCellPrimaryItem_unstable(props, ref);

  useTableCellPrimaryItemStyles_unstable(state);
  return renderTableCellPrimaryItem_unstable(state);
});

TableCellPrimaryItem.displayName = 'TableCellPrimaryItem';
