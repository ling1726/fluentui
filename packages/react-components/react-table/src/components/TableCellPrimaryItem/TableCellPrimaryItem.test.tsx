import * as React from 'react';
import { render } from '@testing-library/react';
import { TableCellPrimaryItem } from './TableCellPrimaryItem';
import { isConformant } from '../../common/isConformant';
import { TableCellPrimaryItemProps } from './TableCellPrimaryItem.types';
import { tableContextDefaultValue, TableContextProvider } from '../../contexts/tableContext';

const tr = document.createElement('tr');
describe('TableCellPrimaryItem', () => {
  beforeEach(() => {
    document.body.appendChild(tr);
  });
  isConformant({
    Component: TableCellPrimaryItem as React.FC<TableCellPrimaryItemProps>,
    displayName: 'TableCellPrimaryItem',
    renderOptions: {
      container: tr,
    },
    testOptions: {
      'has-static-classnames': [
        {
          props: {
            media: 'Test Icon',
            main: 'Main content',
            secondary: 'Secondary content',
          },
        },
      ],
    },
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<TableCellPrimaryItem>Default TableCellPrimaryItem</TableCellPrimaryItem>, { container: tr });
    expect(result.container).toMatchSnapshot();
  });

  it('renders as div if `noNativeElements` is set', () => {
    const { container } = render(
      <TableContextProvider value={{ ...tableContextDefaultValue, noNativeElements: true }}>
        <TableCellPrimaryItem>Table cell</TableCellPrimaryItem>
      </TableContextProvider>,
    );
    expect(container.firstElementChild?.tagName).toEqual('DIV');
    expect(container.firstElementChild?.getAttribute('role')).toEqual('cell');
  });
});
