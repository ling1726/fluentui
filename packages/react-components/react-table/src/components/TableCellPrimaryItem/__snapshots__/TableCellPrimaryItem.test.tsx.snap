// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCellPrimaryItem renders a default state 1`] = `
<tr>
  <div
    class="fui-TableCellPrimaryItem fui-TableCellItem"
  >
    Default TableCellPrimaryItem
  </div>
</tr>
`;
